### ISSUE TITLE
X509: wrongly rejects cert without CDP if CRL contains IDP with name matching cert issuer

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#23325]
When a cert without a CDP extension is validated with CRL checking required
where a CRL from the same issuer is available having an IDP extension, 
the cert is wrongly rejected (with a maybe misleading error: `X509_V_ERR_DIFFERENT_CRL_SCOPE`).

With the certs and CRLs in [demoCA.zip](https://github.com/openssl/openssl/files/13968052/demoCA.zip):
```
openssl verify -attime 1705520210 -trusted cacert.pem -crl_check -CRLfile crl-no-IDP.pem cert-no-CDP.pem 
cert-no-CDP.pem: OK
```
```
openssl verify -attime 1705520210 -trusted cacert.pem -crl_check -CRLfile crl-no-IDP.pem cert-with-CDP.pem
cert-with-CDP.pem: OK
```
```
openssl verify -attime 1705520210 -trusted cacert.pem -crl_check -CRLfile crl-with-IDP.pem cert-with-CDP.pem
cert-with-CDP.pem: OK
```
```
openssl verify -attime 1705520210 -trusted cacert.pem -crl_check -CRLfile crl-with-IDP.pem cert-no-CDP.pem
CN=x509-check-test
error 44 at 0 depth lookup: different CRL scope
error cert-no-CDP.pem: verification failed
```

[RFC 5280 section 6.3.3: CRL Processing](https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3) essentially says that when checking a cert,
for each CDP extension in the cert and for each corresponding CRL, if the CRL has the IDP extension, the IDP must match.
Yet what `get_crl_score()` and `crl_crldp_check()` in `x509_vfy.c` (roughly speaking) do is:
for each available CRL that has the IDP extension with a distpoint (or has a different issuer), there must be a matching CDP extension in the cert.

BTW, test coverage for CRL use is insufficient.

@vdukhovni, this is the bug I mentioned in https://github.com/openssl/openssl/pull/23043#discussion_r1455039798.

**Comment by DDvO** [comment]
Hmm, I meanwhile doubt myself that this is a bug - rather more likely it is intended behaviour, 
since I just found in [RFC 5280 section 5.2.5: Issuing Distribution Point] (https://datatracker.ietf.org/doc/html/rfc5280#section-5.2.5) this sentence:

> If the distributionPoint field is
>  present, then it MUST include at least one of names from the
>  corresponding distributionPoint field of the cRLDistributionPoints
> extension of every certificate that is within the scope of this CRL.

**Comment by DDvO** [comment]
In the light of RFC 5280 section 5.2.5, section 6.3.3 may be misleading (or possibly even contradictory).

@vdukhovni and/or others, what do you think?

**Comment by DDvO** [comment]
BTW, looks like in `crl_crldp_check()` the following check from https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3 is missing:

                  If the distribution point name is present in the IDP CRL
                  extension and the distribution field is omitted from
                  the DP, then verify that one of the names in the IDP
                  matches one of the names in the cRLIssuer field of the DP.

**Comment by dcooper16** [comment]
The relevant text for the situation above comes at the end of [Section 6.3.3](https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3):

> If [after checking every distribution point (DP) in the certificate's CRL distribution points extension] the revocation status has not been determined, repeat the process above with any available CRLs not specified in a distribution point but issued by the certificate issuer.  For the processing of such a CRL, assume a DP with both the reasons and the cRLIssuer fields omitted and a distribution point name of the certificate issuer.  That is, the sequence of names in fullName is generated from the certificate issuer field as well as the certificate issuerAltName extension.  After processing such CRLs, if the revocation status has still not been determined, then return the cert_status UNDETERMINED.

So, in the above case in which there is no CRL distribution points extension, the process in Section 6.3.3 would immediately skip to this step, and then processing of the CRL would fail at step (b)(2)(i) since "the distribution point name is present in the IDP CRL extension," but none of the names in the IDP appear in the certificate's issuer field or issuerAltName extension.

**Comment by DDvO** [comment]
Thank you @dcooper16 for pointing this out!
I did not notice in RFC 5280 the paragraph you mentioned.
Your response explains well why indeed the cert needs be rejected as no suitably scoped CRL can be found:
In the example `crl-with-IDP.pem`, the distribution point name is present in the IDP CRL extension, namely
```
    CRL extensions:
        ...
        X509v3 Issuing Distribution Point: 
            Full Name:
              URI:http://test/ca.crl
```
but that URI does not match the `crl-with-IDP.pem` certificate's issuer field (which is: `C=AU, O=Dodgy Brothers, CN=Dodgy CA`) nor any issuerAltName extensions (since these are not present in the cert).

Nevertheless, the OpenSSL code does not correctly implement the scope checks 
in case an IDP is present in a CRL but no CDPs are present in the target cert:
The actual reason why it rejects the CRL is that `crl_crldp_check()` cannot find a matching CDP (as there is none),
while according to [Section 6.3.3](https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3), as you pointed out, the correct reason for rejection would be
that the cert issuer name does not match any name given in the IDP CRL extension.

In other words, `crl_crldp_check()` in `x509_vfy.c` is incomplete because it does not perform the checks required by the last paragraph of RFC 5280 [section 6.3.3](https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3).

**Comment by t8m** [comment]
I believe this was fixed with #29018

**Comment by DDvO** [comment]
> I believe this was fixed with [#29018](https://github.com/openssl/openssl/pull/29018)

Not really - that one concerned the Indirect CRL flag.

The remaining issue here is that, as I explained [above](https://github.com/openssl/openssl/issues/23325#issuecomment-1921901704),
in case an IDP is present in a CRL but no CDPs are present in the target cert, the CRL gets rejected wrongly (or for a wrong reason) because the CRL scope matching implementation is incomplete: 
the CRL must be used (i.e., its scope accepted) also if any of the names in its IDP matches the issuer (or issuer alt name) of the cert.

I have just created an updated [demoCA.zip](https://github.com/user-attachments/files/24679472/demoCA.zip) to demonstrate this:

```shell
openssl verify -attime 1768585400 -trusted demoCA/cacert2.pem -crl_check -CRLfile demoCA/crl-with-macthing-IDP.pem demoCA/cert2-no-CDP.pem
```
should succeed because the name in the IDP extension of the given CRL: `C=AU, O=Dodgy Brothers, CN=Dodgy CA` does match the cert issuer.
Yet verification wrongly fails with
```
CN=x509-check-test2
error 44 at 0 depth lookup: different CRL scope
error demoCA/cert2-no-CDP.pem: verification failed
```

Some info for reproducing such a CRL:
I used
```shell
openssl ca -gencrl -crlexts crl_ext >demoCA/crl-with-macthing-IDP.pem
```
where the config file contains
```
[ crl_ext ]

# CRL extensions.
# Only issuerAltName and authorityKeyIdentifier make any sense in a CRL.

# issuerAltName=issuer:copy
authorityKeyIdentifier=keyid:always
issuingDistributionPoint=@crl_idp

[ crl_idp ]
# fullname=URI:http://test/ca.crl
fullname=dirName:crl_DN

[ crl_DN ]
C="AU"
O="Dodgy Brothers"
CN="Dodgy CA"
```

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in fixing the bug or if you have reasons why this bug is important enough so we should consider implementing it with higher priority than other requests. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.

**Comment by DDvO** [comment]
Re-opening since community help has arrived:
#32553 provides a fix for the issue.
